pyne/vta.py

```diff
@@ -0,0 +1,481 @@
+    t_min = min(i for i in t_maxs if i > 0)
+    update_dir = list(t_maxs).index(t_min)
+    update_idx = [0, 0, 0]
+    update_idx[update_dir] = 1
+    idxs = [x + y for x, y in zip(idxs, update_idx)] 
+    tp += t_min * v
+    t_temp += t_min
+    return idxs, tp, t_temp
+
+def _move_to_boundary(tp, t_min, v):
+    """
+    Move the point to the nearest boundary when initializing the problem.
+
+    Parameters:
+    -----------
+    tp : numpy array
+        An array of size 3. Temporary point. Actually, it is also the start point.
+    t_min : float
+        Distance to the nearest (positive) boundary. For a 3D mesh, there are 6
+        outside boundaries. Therefore there are 6 value of the distance from the
+        start to the boundaries. This t_min is the smallest one among the
+        positive distance. (The distance is negtive when the direction is
+        negtive.)
+    v : numpy array
+        Direction vector.
+
+    Return:
+    -------
+    tp : numpy array
+        Modified point. This point lies on the nearest boundary. Therefore, we
+        moved point from outside the mesh to the mesh boundary (or the extension
+        of the boundary, which still outside the mesh).
+    """
+
+    tp += t_min * v
+    return tp
+
+def _ray_voxel_traverse(bounds, start, end):
+    """
+    Voxel traversal algorithm.
+    Return the voxels that intersect with the line segment.
+
+    Parameters:
+    -----------
+    bounds : list
+        Boundaries of the mesh grid.
+    start : numpy array
+        Start point.
+    end : numpy array
+        End point.
+
+    Return:
+    -------
+    voxles : list of tuples
+        List of voxel indices. These voxels intersect with the line segment.
+    """
+    # Initialization
+    # Find the voxel that start point in
+    tp = np.copy(start)
+    idxs = [-1, -1, -1]
+    voxels = set()
+    v = _cal_dir_v(start, end)
+    t_end = _distance(start, end)
+    t_temp = 0.0
+    while t_temp < t_end:
+        if _is_point_in_mesh(bounds, tp):
+            # the point is in the mesh
+            # calculate the voxel id
+            for dr in range(len(idxs)):
+                idxs[dr] = _find_voxel_idx_1d(bounds[dr], tp[dr], v[dr])
+            if -1 in idxs:
+                # point outside the mesh
+                return voxels
+            # add current voxel
+            voxels.add((idxs[0], idxs[1], idxs[2]))
+            t_maxs = _cal_max_travel_length_in_current_voxel(tp, end, v, bounds)
+            idxs, tp, t_temp = _move_to_next_voxel(idxs, tp, t_temp, t_maxs, v)
+        else:
+            # Check whether the ray intersecs the mesh. Calculate the maxinum travel
+            # length of the ray to the outside boundaries of the mesh. If all the
+            # length are negtive, then there is no intersection, return. Ohterwise,
+            # choose the minimum length and move the point to that position.
+            # left, right, back, front, down, up
+            t_maxs = [0.0] * 6
```

Comment:
The code here is kind of what I'd expect to see in the `_move_to_boundary` function defined above.

Ideally I think there would be two calls here:

```
intersects = _check_mesh_intersection(start, end, bounds) # check if they ray intersects the mesh
if not intersects:
    break
pnt = _move_to_boundary(pnt, bounds)
```

or something like that.

Reply:
To me, it's less about the number of lines and more about separating the functionality, which you've done really well in other places in this module.

Reply:
To put it another way, there are many functions in this module which are only a few lines, but they serve a distinct purpose and make the code easy to follow when used in other places. I think that practice could be repeated here.

Reply:
I used the `_move_to_boundary` here in the new commit. 
As for the intersection checking function that checks whether the line segment intersects with the mesh is not finished by a single function. This check is finished during the `while` loop. In each single the loop step, the if clause below can't tell whether there will be intersection if it's ture, but there will definitely no intersection if it's false. 
```
if any((x > 0).all() for x in dist_maxs):
```
So, this is not an independent checking function.